designs/dotnetcore-debugging.md

```diff
@@ -0,0 +1,196 @@
+
+### Success Criteria
+
+1. Provide handy script to get the `vsdbg` via downloading it to the .NET Lambda runtime container, installing it and getting through the mounting (or manual instructions to achieve the above results). After that user would be able to provide it to SAM via `--debugger-path`;
+2. Provide ready-to-go `launch.json` configuration for attaching to the Lambda function via VSCode;
+3. Customer should be able to easily debug .NET Core 2.0 and 2.1 apps on VS Code (minimal requirement).
+
+### Out-of-Scope
+
+Visual Studio 2017 (can be a target, but it has some other very different Docker container debugging support - needs additional investigation). Help in this area is highly appreciated.
+
+### User Experience Walkthrough
+
+##### 1. Getting the debugger locally
+
+At first, user should be able to get and install `vsdbg` on their machine easily. For this we should provide the script for them to invoke, which will effectively run the .NET Lambda runtime Docker container with mounted via `bind` path to get the debugger on the **host** machine. In this container following: `curl -sSL https://aka.ms/getvsdbgsh | bash /dev/stdin -v latest -l /vsdbg`  should be run to get and install required debugger (specifically built for this container). Later SAM will mount this same folder with the debugger to the target running container. 
+
+Instead of script we can certainly provide manual instructions of how to do the above stuff step-by-step as it is now done for golang debugger. _Feedback appreciated_.
+
+Customer should go through this step **only once** to set up the debugger.
+
+##### 2. Publishing the code
+
+Customer should be informed, that in order to have the best debugging experience app must be published in `Debug` configuration by supplying `-c Debug` flag for `dotnet publish` command. 
+
+_For example:_
+
+```sh
+dotnet publish -c Debug -o out
+
+# Or via container
+docker run --rm --mount src=$(pwd),dst=/var/task,type=bind lambci/lambda:build-dotnetcore2.1 dotnet publish -c Debug -o out
+```
+
+##### 3. Attaching
+
+Customer will use ready-to-go `launch.json` debug configuration for VS Code, fill in the container name for the `docker exec` and supply this name to SAM.
+
+_For example:_
+
+```sh
+sam local invoke --event event.json -d 6000 --debugger-path ~/vsdbg --container-name debugger "MyFunction"
```

Comment:
will the codeuri in the template.yaml be any different, if the app is published under debug mode?